### ISSUE TITLE
How to query for other fields in a resolver before making http request?

### ISSUE THREAD

**Issue by sgrove** [andreas/ocaml-graphql-server#42]
I'm working with a REST backend that allows you to specify in the query params the fields you want along with some arguments about formatting. The problem right now is that when I enter that when I enter the top node, I have an `io_field` where I have to make the request, and then I pass on the result to the lower fields - but I actually want to know what those fields are so that I can make the appropriate request (rather than pulling out the very large full record all the time).

A few examples of a query and the url it would be translated into when resolving:

```
query {
  tweet(id:"123") {
    username
  }
}
```

`myapi.com/query?fields=username`

```
query {
  tweet(id:"123") {
    username
    body
  }
}
```

`myapi.com/query?fields=username,body`

```
query {
  tweet(id:"123") {
    username
    body(limit: 100)
  }
}
```

`myapi.com/query?fields=username,body[limit:100]`

In the twitter example size isn't such a big deal, but imagine lots of multi-megabyte fields in the db that I'd have to pull over by default right now when resolving `tweet` just in case the `reallyReallyBigField` was used, and also that there are *lots* of these kinds of fields so making different `tweetWithReallyReallyBigField { ... }` isn't really any option.

Any chance the resolver can be passed some context about its immediate children in a way that makes this easy to do?

**Comment by andreas** [comment]
This seems like an entirely sane use case, and `graphql-js` also exposes a [`GraphqlResolveInfo`](https://github.com/graphql/graphql-js/blob/3c0a536688ae53df362c9777170e91aa66114760/src/type/definition.js#L598-L609)-object to resolvers:

```javascript
export type GraphQLResolveInfo = {
  fieldName: string;
  fieldNodes: Array<FieldNode>;
  returnType: GraphQLOutputType;
  parentType: GraphQLCompositeType;
  path: ResponsePath;
  schema: GraphQLSchema;
  fragments: { [fragmentName: string]: FragmentDefinitionNode };
  rootValue: mixed;
  operation: OperationDefinitionNode;
  variableValues: { [variableName: string]: mixed };
};
```

Is it something like that you had in mind?

The simplest solution would be to add an argument like this to all resolve functions. On the other hand I don't really like forcing all all resolve functions to receive this argument, when I expect the vast majority will not use it.

A solution which addresses this would be to introduce a separate function, say `field'`, for fields where the resolve function expects a resolve info argument:

```ocaml
val field' : ?doc:string ->
             string ->
             typ:('ctx, 'a) typ ->
             args:('a, 'b) Arg.arg_list ->
             resolve:(resolve_info -> 'ctx -> 'src -> 'b) -> (* <-- resolve_info is new *)
             ('ctx, 'src) field
```

A third approach would be to have a special "arg type", which indicates that the resolve function expects a resolve info argument:

```ocaml
let query = Graphql.(obj
  "query"
  ~fields:(fun _ -> [
    field "tweet"
      ~typ:tweet
      ~args:Arg.[
        arg "id" ~typ:(non_null int);
        resolve_info (* <-- indicates resolve function expects resolve information *)
      ]
      ~resolve:(fun id resolve_info ->
        (* ... *)
      )
  ])
)
```

This solution has some unfortunate implications though: you can request resolve info multiple times, the resolve info argument needs to filtered out for introspection, etc.

Any thoughts on the API, @sgrove ?

**Comment by sgrove** [comment]
I like these approaches - two other ideas:

Following your design above, if their prep work before the query was another optional argument, then memoizing it would be easy (and it seems like it should be a memoizable function)
```ml
let query = Graphql.(obj
  "query"
  ~fields:(fun _ -> [
    field "tweet"
      ~typ:tweet
      ~args:Arg.[
        arg "id" ~typ:(non_null int);
        resolve_info (* <-- indicates resolve function expects resolve information *)
      ]
      (* Use prep to take in args + resolve info and output some data
         to be used by resolve, e.g. finding fields. Can memoize
         results of this function easily. *)
      ~prep:(fun id resolve_info : string list -> find_fields resolve_info)
      (* list_of_fields obviously just an example, might be other
         contextual info needed about the overall query at the time of
         resolution *)
      ~resolve:(fun id list_of_fields ->
        (* ... *)
      )
  ])
)
```

Alternatively, simply having a different key instead of `~resolve`, `~resolve_info` that had the extra baggage that you mentioned could keep the current simpler one the same.

Honestly, I'm happy to try either way, whatever is the fastest thing for you to implement and for us to try out!

Would you mind sketching out what pulling out the different fields would look like with the type you have in mind right now? It wasn't immediately obvious to me from the `GraphQLResolveInfo` above, and just want to make sure it would cover our use case before spending time on it.

**Comment by sgrove** [comment]
We're bumping up against this a bit, have you had time to think through the design any further?

**Comment by andreas** [comment]
Thanks for bringing this up again. Inspired by `graphql-go`, I had this idea of changing the type of resolve functions to `'ctx resolve_params -> 'src -> 'a`, where `resolve_params` is defined as:

```ocaml
type 'ctx resolve_params = {
  ctx: 'ctx;
  info: resolve_info (* as in previous comment -- could possible be lazy *)
}
```

So this would change the type of the first argument of a resolve function to become a record that contains both the context, query metadata and possibly other information. A resolve function would then look like `fun params account arg1 -> params.ctx.fetchOwner account.ownerId arg1`. It's slightly more clumsy than having straight access to the context though, but it avoids adding yet another argument to all resolve functions.

**Comment by andreas** [comment]
@sgrove I've implemented the `resolve_params`-idea here: https://github.com/andreas/ocaml-graphql-server/tree/resolve-params

Give it a spin -- hopefully it can unblock you. I don't feel confident about the API though, so it may be subject to change.

**Comment by sgrove** [comment]
This actually looks pretty good for us, but looking at https://github.com/andreas/ocaml-graphql-server/commit/da5086f28b37f9a319ce87a6e075a2f18842f278#diff-374482d3d06ebc97235b7becd2cfb4a5R296, would that field allow us to

**Comment by andreas** [comment]
Yes, you can descend further down. The simple API I cooked up here doesn't allow you to resolve fragments though...

**Comment by osener** [comment]
I'd love to have this feature as well!
